### PR TITLE
Probable fix for out of place BinaryOpScalar bad values and/or IMAs on 11.2

### LINKED ISSUE #51992: test_optim.py failures for CUDA 11.2 on CI
## 🐛 Bug

The following tests in test_optim.py fail in CI on CUDA 11.2 in linux and windows.
test_adadelta
test_adam
test_adamw
test_multi_tensor_optimizers
test_rmsprop

## To Reproduce

1. Run the CI jobs pytorch_linux_xenial_cuda11_2_cudnn8_py3_gcc7_test and pytorch_windows_vs2019_py36_cuda11.2_test2
2. Get the following logs with error messages:
Linux: https://app.circleci.com/pipelines/github/pytorch/pytorch/271291/workflows/ffe1be08-5d55-459e-98bd-200d6387e41c/jobs/10779915
Windows: https://app.circleci.com/pipelines/github/pytorch/pytorch/271865/workflows/2ac5bd63-6dba-4169-9f5d-7ccaf7ab33b4/jobs/10806271

## Expected behavior

Those tests should pass.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @seemethere @malfet @walterddr @pytorch/pytorch-dev-infra @ngimel

### PR DESCRIPTION
**Pull request by mcarilli**
Should close https://github.com/pytorch/pytorch/issues/51992.

Suggested by one of our compiler people (Hari Sandanagobalane, don't know github handle). Also big thanks to @ngimel @zasdfgbnm for distilling a minimal repro of the original failures.

Good news is, the bug is a likely a foreach kernel bug and not an 11.2 compiler bug. Therefore, in theory it could affect any cuda version. We think 11.2 exposed it by optimizing more aggressively than previous toolkits.

IIRC some foreach tests were disabled because of this bug, but I'm not sure which or how many. @ngimel @izdeby what tests should I reenable?

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
Please reenable these tests: https://github.com/pytorch/pytorch/pull/51598/files#diff-f2b37d0b5812153acf9ff1e337d375c5cad0076c4ddc30535ec2d55c8ac9b770

**Review by izdeby — APPROVED** [review]

**Comment by izdeby** [comment]
@mcarilli, thank you for this! Looks like the [fix is working](https://github.com/pytorch/pytorch/pull/52602).

**Comment by ngimel** [comment]
Please change tests https://github.com/pytorch/pytorch/blob/4386a3803c707d0ef1bcf29005ae45cfd2f923e8/test/test_foreach.py#L353 to make sure that this is caught by foreach tests (it definitely was not caught previously, foreach tests were passing on 11.2)

File: aten/src/ATen/native/cuda/ForeachFunctors.cuh [hunk 1]
```diff
@@ -99,7 +99,8 @@ __device__ __forceinline__ void binary_op_scalar(
         }
         else {
             for(int i_start = 0; i_start < n && i_start < chunk_size; i_start += blockDim.x * kILP) {
-                load_args<depth>(r_args, args, i_start, chunk_size, n);
+                // Regardless if "depth" is 1 (for inplace) or 2 (for out of place), r_args has depth 1
+                load_args<1>(r_args, args, i_start, chunk_size, n);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/ForeachFunctors.cuh:103]: `depth` template argument is no longer needed

**Comment by mcarilli** [comment]
Closing to resubmit to ci-all (https://github.com/pytorch/pytorch/pull/52634).

**Pull request closed without merging**